### PR TITLE
Py mesa

### PR DESCRIPTION
**Pull request by jdmuss**
Beginning of PyMesa qucikstart project. Python code to mimic the quickstart Accumulo examples. Good through queries. Working on insert code.

### CODE REVIEW

File: geomesa-quickstart-python/geomesa_quickstart_jnius.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
+#coding:utf-8
+"""
+geomesa_quickstart_jnius.py
+
+Note: this has been tested with python 2.7.12 & 3.5
+
+Description:
+   This script is a python adaptation of the GeoMesa accumulo quickstart. It is a querry only
```
Comment (aheyne) [inline geomesa-quickstart-python/geomesa_quickstart_jnius.py]: [Picky] Capitalize Accumulo

Comment (aheyne) [reply geomesa-quickstart-python/geomesa_quickstart_jnius.py]: 's/querry/query/'

File: geomesa-quickstart-python/geomesa_quickstart_jnius.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+   demo, in contrast to the java quickstart that creates the table & loads data into it. This
+   script uses two second party packages (jnius & bokeh) that need to be loaded to run. 'jnius'
```
Comment (aheyne) [inline geomesa-quickstart-python/geomesa_quickstart_jnius.py]: I'd copy the styling JNIus and Bokeh use unless it's a command. (other places too)

File: geomesa-quickstart-python/geomesa_quickstart_jnius.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+How to use:
+Updates:
+
+To Do: Get queryFeaturesToDict from tools/ECQL/queryFormatters
+"""
+from __future__ import print_function
+from datetime import datetime
+'''----------------------------------------------------------------------------------------------------------------------'''
+''' Import Java & Geomesa classes: '''
```
Comment (aheyne) [inline geomesa-quickstart-python/geomesa_quickstart_jnius.py]: s/Geomesa/GeoMesa/

File: geomesa-quickstart-python/geomesa_quickstart_jnius.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+'''----------------------------------------------------------------------------------------------------------------------'''
+from utils.geomesa_jnius_setup import *
+from utils.quickstart_command_line_parser import getArgs
+from pyJavaClasses.datastore import getDataStore as getDataStore, createAccumuloDBConf
```
Comment (aheyne) [inline geomesa-quickstart-python/geomesa_quickstart_jnius.py]: Is the alias needed?

File: geomesa-quickstart-python/pyJavaClasses/FeatureTypes.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        self.simpleFeature = None
+        self.true = getJavaTrueFalse(JNI, True)
+    
+    def createSimpleFeatureType(self, simpleFeatureTypeName, attribute_list):
+        """ This will create a 'bare' simpleFeatureType from a list of attributes: """
+        attributes = ','.join(attribute_list)
+        self.simpleFeature = self.SimpleFeatureTypes.createType(simpleFeatureTypeName, attributes)
+    
+    def setDateTimeIndex(self, field_name):
+        """ use the user-data (hints) to specify which date-time field is meant to be indexed: """
```
Comment (aheyne) [inline geomesa-quickstart-python/pyJavaClasses/FeatureTypes.py]: [Picky] Capital ``u``

File: geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+To Do:
+"""
+class simpleFeatureType():
+    def __init__(self, JNI):
+        """ Create the necessary java class interfaces: """
+        self.SimpleFeatureTypes = JNI.autoclass("org.locationtech.geomesa.utils.interop.SimpleFeatureTypes")
+
+    def createSimpleFeatureType(self, simpleFeatureTypeName,  attribute_list):
+        """ This will create a 'bare' simpleFeatureType from a list of attributes: """
```
Comment (aheyne) [inline geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py:33]: What's a bare sft?

File: geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        attributes = ','.join(attribute_list)
+        simpleFeatureType = self.SimpleFeatureTypes.createType(simpleFeatureTypeName, attributes)
+        return simpleFeatureType
+
+    def setDateTimeIndex(self, simpleFeature, field_name):
+        """ use the user-data (hints) to specify which date-time field is meant to be indexed: """
+        simpleFeature.getUserData().put(self.SimpleFeatureTypes.DEFAULT_DATE_KEY, field_name);
```
Comment (aheyne) [inline geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py]: Semicolon?

File: geomesa-quickstart-python/pyJavaClasses/customJavaClasses.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    class Boolean(JNI.JavaClass):
+        __javaclass__ = 'java/lang/Object'
+        __metaclass__ = JNI.MetaJavaClass
+        TRUE = JNI.JavaObject('()Zjava/lang/Object;', static=True)
+        FALSE = JNI.JavaObject('()Zjava/lang/Object;', static=True)
+    
+    if isinstance(pyBool, str):
+        boolVal = BoolDict.get(pyBool.lower(), None)
+        if not isinstance(boolVal, bool):
+            print("Error:bad balue passed to 'getJavaTrueFalse'; {} must be in str(true, false, yes, no).".format(pyBool))
```
Comment (aheyne) [inline geomesa-quickstart-python/pyJavaClasses/customJavaClasses.py]: s/balue/value/

File: geomesa-quickstart-python/pyJavaClasses/customJavaClasses.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+            sys.exit(-1)
+    elif isinstance(pyBool, bool):
+        boolVal = pyBool
+    else:
+        print("Error:bad balue passed to 'getJavaTrueFalse'; {} must be type bool or in str(true, false, yes, no).".format(pyBool))
```
Comment (aheyne) [inline geomesa-quickstart-python/pyJavaClasses/customJavaClasses.py]: s/balue/value/

File: geomesa-quickstart-python/pyJavaClasses/javaDateTime.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python
+#coding:utf-8
+"""
+javaDateTime.py
+
+Description:
+   This function takes a python datetime object with optional timezone information, translates
+   it into a java datetime object ( using jnius) and returns the java datetime object.
```
Comment (aheyne) [inline geomesa-quickstart-python/pyJavaClasses/javaDateTime.py]: [picky] spacing in ``()``

File: geomesa-quickstart-python/tools/ECQL/filter.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+class ECQLQuery:
+    """ This is a helper class to allow python access to java Query & CQL functions."""
+    def __init__(self, JNI):
+        ''' This is sets up links to the java types & java Query and CQL functions.'''
+        self.jString = JNI.autoclass('java.lang.String')
+        self.Query = JNI.autoclass('org.geotools.data.Query')
+        self.CQL = JNI.autoclass('org.geotools.filter.text.cql2.CQL')
+    
+    def createFilter(self, filter_string):
+        ''' Create the (E)CQL filter from the prepared (python) string: '''
```
Comment (aheyne) [inline geomesa-quickstart-python/tools/ECQL/filter.py]: Should probably standardize to ``(E)CQL`` or ``ECQL``.

File: geomesa-quickstart-python/tools/ECQL/queryFormatters.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+Creation Date: 4-11-2017
+Version:       1.0
+
+Dependencies: 
+         Public:     jnius, json
+         Private:   ECQL, setupJnius
+
+Interfaces:
+            queryFeaturesToDict         Execute a query, convert the java objects to python types, & load
+                                                     load the results into a python dict of dicts. The master dict is keyed
```
Comment (aheyne) [inline geomesa-quickstart-python/tools/ECQL/queryFormatters.py]: s/& load load/and load/

File: geomesa-quickstart-python/utils/geomesa_command_line_parser.py [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+                       help="the name of your Accumulo instance (default=local)")
+        self.add_argument("--zookeepers", "--zoo", "-z", default="localhost", 
+                       help="your Zookeeper nodes, separated by commas (default=localhost)")
+        self.add_argument("--user", "-u", default="root", 
+                       help="the name of an Accumulo user that has permissions to create, read and write tables (default=root)")
+        self.add_argument("--password", "--pwd", "--pw", default="secret", 
+                       help="the password for the previously-mentioned Accumulo user (default=secret)")
+        self.add_argument("--tableName", "--tableNm", "--table", "--tbl", "-t",  default="quickstart", 
+                       help="the name of the on which to operate. This table will be created if it does not exist (default=quickstart)")
+        self.add_argument("--classpath", "-c", type=str, default=None,
```
Comment (aheyne) [inline geomesa-quickstart-python/utils/geomesa_command_line_parser.py]: ``-c`` could possibly get confused with catalog parameter in the tools. Maybe make this ``-cp``? Not sure if we care, just a thought.

File: geomesa-quickstart-python/utils/geomesa_jnius_setup.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+    @property
+    def detach(self):
+        ''' Detach jnius' native thread to the JVM:'''
+        return _detach
+    
+    def find_javaclass(self, JavaClass, msg=None):
+        if msg is None : msg = JavaClass
+        try:
+            _find_javaclass(JavaClass)
+            print("{} class path is OK".format(msg))
```
Comment (aheyne) [inline geomesa-quickstart-python/utils/geomesa_jnius_setup.py]: s/class path/classpath/

File: geomesa-quickstart-python/utils/geomesa_jnius_setup.py [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+            result = True
+        except:
+            print("Error: {} class path is NOT OK".format(msg))
```
Comment (aheyne) [inline geomesa-quickstart-python/utils/geomesa_jnius_setup.py]: s/class path/classpath/

**Review by aheyne — COMMENTED** [review]
Looks great, I'll try and test it tomorrow. These comments are basically all [picky].

**Comment by aheyne** [comment]
Also going to need copyright headers.

File: geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+Interfaces:
+        createSimpleFeatureType              Class to create a 'bare' simpleFeatureType from a list of attributes,
+                                                              and to perform appropriate tasks, e.g. set feature indices
+
+Updates:
+
+To Do:
+"""
+class simpleFeatureType():
```
Comment (tkunicki) [inline geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py]: What's Python convention for class names? Think this should be `SimpleFeatureType`

File: geomesa-quickstart-python/tools/dbModify/insert.py [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+from __future__ import print_function
+
+def insert(JNI, ecql, simpleFeatureTypeName, dataStore, filter_string, field_dict, print_num=10):
+    ''' Insert data into an accumulo simpleFeatureType (table): '''
+    jLong = JNI.autoclass('java.lang.Long')
+    WKTUtils = JNI.autoclass("org.locationtech.geomesa.utils.interop.WKTUtils")
+    dfc = JNI.autoclass("org.geotools.feature.DefaultFeatureCollection")
+    SimpleFeatureStore = JNI.autoclass("org.geotools.data.simple.SimpleFeatureStore")
+    
+    # ToDo: This is test stuff, delete this block when finished
```
Comment (tkunicki) [inline geomesa-quickstart-python/tools/dbModify/insert.py:35]: This still valid?  Does this code need to move to test?

File: geomesa-quickstart-python/utils/geomesa_jnius_setup.py [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+
+To Do:
+"""
+'''----------------------------------------------------------------------------------------------------------------------'''
+''' Import Java & Geomesa classes: '''
+'''----------------------------------------------------------------------------------------------------------------------'''
+#from __future__ import print_function
+import os, sys
+
+class setupJnius:
```
Comment (tkunicki) [inline geomesa-quickstart-python/utils/geomesa_jnius_setup.py]: Again, naming per conventions

File: geomesa-quickstart-python/tools/dbModify/insert.py (hunk 17 above)
Comment (jdmuss) [reply geomesa-quickstart-python/tools/dbModify/insert.py:35]: Still working on wrapping the test code. This was just a place holder. Either I need to finish it, or remove it and make a branch of my branch

File: geomesa-quickstart-python/geomesa_quickstart_jnius.py (hunk 2 above)
Comment (jdmuss) [reply geomesa-quickstart-python/geomesa_quickstart_jnius.py]: I don't follow you on this one.

File: geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py (hunk 6 above)
Comment (jdmuss) [reply geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py:33]: No idea. It is in the AccumuloQuickStart demo. In static SimpleFeatureType createSimpleFeatureType(String simpleFeatureTypeName); line 134:// create the bare simple-feature type
Maybe empty simpleFeatureType?

File: geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py (hunk 16 above)
Comment (jdmuss) [reply geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py]: Yeah PEP-8 says Camelcase

File: geomesa-quickstart-python/geomesa_quickstart_jnius.py (hunk 2 above)
Comment (aheyne) [reply geomesa-quickstart-python/geomesa_quickstart_jnius.py]: Like how we style geomesa -> GeoMesa. I think bokeh -> Bokeh and jnius -> JNIus but from looking did they changed the name of JNIus to [PyJNIus](https://github.com/kivy/pyjnius)?

File: geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py (hunk 16 above)
Comment (tkunicki) [reply geomesa-quickstart-python/pyJavaClasses/SimpleFeatureType.py]: PEP-8 says [CapWords for class names](https://www.python.org/dev/peps/pep-0008/#class-names)

**Pull request merged**